### PR TITLE
Add support for smooth pinch-to-zoom on a trackpad (bug 1659492)

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

**Comment by calixteman** [comment]
I tested the feature on a windows laptop and on a macbook and in both cases it works well.

File: web/app.js [hunk 1]
```diff
@@ -2651,50 +2654,62 @@ function webViewerWheel(evt) {
   ) {
     // Only zoom the pages, not the entire viewer.
     evt.preventDefault();
+
```
Comment (Snuffleupagus) [inline web/app.js:2657]: Nit: This looks like an unrelated change.

File: web/app.js [hunk 2]
```diff
@@ -2651,50 +2654,62 @@ function webViewerWheel(evt) {
...
-    const deltaMode = evt.deltaMode;
-    const delta = normalizeWheelEventDirection(evt);
     const previousScale = pdfViewer.currentScale;
+    if (isPinchToZoom && supportsPinchToZoom) {
+      // The following formula is a bit strange but it comes from:
+      // https://searchfox.org/mozilla-central/rev/d62c4c4d5547064487006a1506287da394b64724/widget/InputData.cpp#618-626
+      const scaleFactor = Math.exp(-evt.deltaY / 100);
+      if (scaleFactor < 1) {
+        PDFViewerApplication.zoomOut(null, scaleFactor);
+      } else {
```
Comment (Snuffleupagus) [inline web/app.js:2670]: Given how the exponentiation works, it's theoretically possible that `scaleFactor === 1` if `evt.deltaY = 0` ever happens, in which case we'd get into trouble when trying to zoom. Hence you'll need the following here:
```suggestion
      const scaleFactor = Math.exp(-evt.deltaY / 100);
      if (scaleFactor < 1) {
        PDFViewerApplication.zoomOut(null, scaleFactor);
      } else if (scaleFactor > 1) {
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, with the comments addressed; thank you!

File: web/app.js [hunk 3]
```diff
@@ -2656,45 +2659,56 @@ function webViewerWheel(evt) {
...
-    // It is important that we query deltaMode before delta{X,Y}, so that
-    // Firefox doesn't switch to DOM_DELTA_PIXEL mode for compat with other
-    // browsers, see https://bugzilla.mozilla.org/show_bug.cgi?id=1392460.
-    const deltaMode = evt.deltaMode;
-    const delta = normalizeWheelEventDirection(evt);
     const previousScale = pdfViewer.currentScale;
+    if (isPinchToZoom && supportsPinchToZoom) {
+      // The following formula is a bit strange but it comes from:
+      // https://searchfox.org/mozilla-central/rev/d62c4c4d5547064487006a1506287da394b64724/widget/InputData.cpp#618-626
+      const scaleFactor = Math.exp(-evt.deltaY / 100);
```
Comment (Snuffleupagus) [inline web/app.js:2666]: Although, isn't there a risk that this can be too small to be meaningful; do we perhaps want a check similar to this one as well?
https://github.com/mozilla/pdf.js/blob/d6f63b5fb7cc63fa5d38ff71a56b8ee46d4f1b33/web/app.js#L2753-L2756

**Comment by calixteman** [comment]
@Snuffleupagus could you re-review it please ?
I made some changes to take into account your last comment which led me to think that the 1e-2 was not a good threshold (1% of 10% is not the same as 1% of 500%).
I tested successfully those changes on touchscreen and on two laptops (mac and win) with a trackpad.

File: web/app.js [hunk 4]
```diff
@@ -2061,6 +2063,22 @@ const PDFViewerApplication = {
     return wholeTicks;
   },
 
+  _accumulateFactor(previousFactor, factor, prop) {
```
Comment (Snuffleupagus) [inline web/app.js:2066]: Given what you're passing to this function, shouldn't the first parameter be named `previousScale` instead?

File: web/app.js [hunk 5]
```diff
@@ -2743,42 +2783,94 @@ function webViewerTouchMove(evt) {
...
+    const v2X = page0X - page1X;
+    const v2Y = page0Y - page1Y;
+    const det = v1X * v2Y - v1Y * v2X;
+    if (Math.abs(det) > 0.02 * Math.hypot(v1X, v1Y) * Math.hypot(v2X, v2Y)) {
+      return;
+    }
+    pinch = true;
+  }
+
+  if (!pinch) {
```
Comment (Snuffleupagus) [inline web/app.js:2842]: Do we really need the `pinch` variable, since it seems that we could just do the following?
```suggestion
  } else {
```

**Review by Snuffleupagus — COMMENTED** [review]
I'll try to look at this in more detail tomorrow, since there's been a fair number of changes since I reviewed this previously.

File: web/app.js [hunk 6]
```diff
@@ -2061,6 +2063,22 @@ const PDFViewerApplication = {
     return wholeTicks;
   },
 
+  _accumulateFactor(previousScale, factor, prop) {
+    if (factor === 1) {
+      return 1;
+    }
+    // If the direction changed, reset the accumulated ticks.
```
Comment (Snuffleupagus) [inline web/app.js:2070]: Too much copy-and-paste in the comment :-)
```suggestion
    // If the direction changed, reset the accumulated factor.
```

File: web/app.js [hunk 7]
```diff
@@ -2061,6 +2063,22 @@ const PDFViewerApplication = {
...
+    if ((this[prop] > 1 && factor < 1) || (this[prop] < 1 && factor > 1)) {
+      this[prop] = 1;
+    }
+
+    const newFactor = previousScale * factor * this[prop];
+    const wholeFactor = Math.floor(newFactor * 100) / (100 * previousScale);
+    this[prop] = factor / wholeFactor;
+
+    return wholeFactor;
```
Comment (Snuffleupagus) [inline web/app.js:2079]: The `wholeFactor` variable name feels slightly misleading, since it doesn't normally contain a whole number in this method (as opposed to the method above, from which I guess this was copied). How about we simply do the following?
```suggestion
    const newFactor =
      Math.floor(100 * previousScale * factor * this[prop]) / (100 * previousScale);
    this[prop] = factor / newFactor;
    return newFactor;
```

**Review by Snuffleupagus — APPROVED** [review]
